src/storage/datacore/restapi/custom/api.pm

```diff
@@ -0,0 +1,153 @@
+# limitations under the License.
+#
+package storage::datacore::restapi::custom::api;
+use strict;
+use warnings;
+use centreon::plugins::http;
+use centreon::plugins::statefile;
+use JSON::XS;
+use centreon::plugins::misc qw(empty);
+
+sub new {
+    my ($class, %options) = @_;
+    if (!defined($options{output})) {
+        print "Class Custom: Need to specify 'output' argument.\n";
+        exit 3;
+    }
+    if (!defined($options{options})) {
+        $options{output}->add_option_msg(short_msg => "Class Custom: Need to specify 'options' argument.");
+        $options{output}->option_exit();
+    }
+    my $self = {};
+    bless $self, $class;
+
+    $options{options}->add_options(arguments => {
+        'hostname:s' => { name => 'hostname' },
+        'port:s'     => { name => 'port', default => 443 },
+        'proto:s'    => { name => 'proto', default => 'https' },
+        'timeout:s'  => { name => 'timeout' },
+        'username:s' => { name => 'username' },
+        'password:s' => { name => 'password' }
+    });
+    $self->{output} = $options{output};
+    $self->{http} = centreon::plugins::http->new(%options, default_backend => 'curl');
+
+    return $self;
+}
+
+sub set_options {
+    my ($self, %options) = @_;
+
+    $self->{option_results} = $options{option_results};
+}
+sub set_defaults {}
+
+# hostname,username and password are required options
+sub check_options {
+    my ($self, %options) = @_;
+    $self->{http}->set_options(%{$self->{option_results}});
+
+    if (centreon::plugins::misc::is_empty($self->{option_results}->{hostname})) {
+        $self->{output}->add_option_msg(short_msg => 'Please set hostname option');
+        $self->{output}->option_exit();
+    }
+    if (centreon::plugins::misc::is_empty($self->{option_results}->{username})) {
+        $self->{output}->add_option_msg(short_msg => 'Please set username option to authenticate against datacore rest api');
+        $self->{output}->option_exit();
+    }
+    if (centreon::plugins::misc::is_empty($self->{option_results}->{password})) {
+        $self->{output}->add_option_msg(short_msg => 'Please set password option to authenticate against datacore rest api');
+        $self->{output}->option_exit();
+    }
+
+}
+# wrapper around centreon::plugins::http::request to add authentication and decode json.
+# output : deserialized json from the api if not error found in http call.
+sub request_api {
+    my ($self, %options) = @_;
+    # datacore api require a ServerHost header with the hostname used to query the api to respond.
+    # authentication is http standard basic auth.
+    my $result = $self->{http}->request(
+        basic       => 1,
+        credentials => 1,
+        header      => [ "ServerHost: $self->{option_results}->{hostname}" ],
+        username    => $self->{option_results}->{username},
+        password    => $self->{option_results}->{password},
+        %options,
+    );
```

Comment:
Here is missing the handling of code return or empty content in result something like this :

```if ($self->{http}->get_code() < 200 || $self->{http}->get_code() >= 300) {```
```if (!defined($content) || $content eq '') {```

Reply:
this is already handled by the http lib, the `unknown_status` argument of request have a default value : 

https://github.com/centreon/centreon-plugins/blob/6a5fb84d73e10093b067d61c252145c2b5ef9d56/src/centreon/plugins/http.pm#L64